learn/chapter1/0-opengl.md

```diff
@@ -0,0 +1,76 @@
+
+```cs
+if (GL_ARB_extension_name)
+{
+    // Do cool new and modern stuff supported by hardware
+}
+else
+{
+    // Extension not supported: do it the old way
+}
+```
+
+With OpenGL version 3.3 we rarely need an extension for most techniques, but wherever it is necessary proper instructions are provided.
+
+## State machine
+
+OpenGL is by itself a large state machine: a collection of variables that define how OpenGL should currently operate. The state of OpenGL is commonly referred to as the OpenGL context. When using OpenGL, we often change its state by setting some options, manipulating some buffers and then render using the current context.
+
+Whenever we tell OpenGL that we now want to draw lines instead of triangles for example, we change the state of OpenGL by changing some context variable that sets how OpenGL should draw. As soon as we changed the state by telling OpenGL it should draw lines, the next drawing commands will now draw lines instead of triangles.
+
+When working in OpenGL we will come across several state-changing functions that change the context and several state-using functions that perform some operations based on the current state of OpenGL. As long as you keep in mind that OpenGL is basically one large state machine, most of its functionality will make more sense.
+
+## Objects
+
+The OpenGL libraries are written in C and allows for many derivations in other languages, but in its core it remains a C-library. Since many of C's language-constructs do not translate that well to other higher-level languages, OpenGL was developed with several abstractions in mind. One of those abstractions are objects in OpenGL.
+
+OpenGL objects aren't like the objects you'd find in an object-oriented language. You can't directly modify them, since they're stored on the graphics card. Instead, you have an integer that represents where the object is on the graphics card, referred to as a handle. You pass this handle to any function that needs an object. It works like this:
+
+```cs
+//Create a variable to store the handle
+int ObjectHandle;
+
+//Create the object
+GL.GenObjects(1, &ObjectHandle);
+
+//Then pass it to a function that uses it
+GL.SomeFunctionThatUsesThisObject(ObjectHandle);
+```
+
+This little piece of code is a workflow you'll frequently see when working in OpenGL. We first create an object and store a reference to it as an id (the real object data is stored behind the scenes). Then we bind the object to the target location of the context (the location of the example window object target is defined as GL_WINDOW_TARGET). Next we set the window options and finally we un-bind the object by setting the current object id of the window target to 0. The options we set are stored in the object referenced by objectId and restored as soon as we bind the object back to GL_WINDOW_TARGET.
```

Comment:
We should review these (and any other) constants in this article.

Author: @devvoid

Reply:
I don't know what this should be changed to, my google-fu has failed me.